### PR TITLE
gh-101100: Only show GitHub check annotations on changed doc paragraphs

### PR DESCRIPTION
**Pull request by CAM-Gerlach**
PR #102513 implemented support for displaying Sphinx warnings on changed files inline in the GitHub UI, which can be a useful aid to authors that they've fixed any defects on the lines they've touched and not introduced new ones.

However, annotations are shown on an entire file, regardless of whether the line the warning was issued on was within or even near those actually modified by the PR. While this has helped motivate large-scale cleanup of many of these warnings, as I commented when this was first deployed it has naturally led to a serious amount of noise, annoyance and confusion for PR authors seeing warnings on lines they had nothing to do with, which in turn leads to the understandable temptation for authors to ignore valid warnings or just silencing them, hiding rather than fixing the underlying defect.

The solution is, naturally,  to only show warnings on modified lines, which is somewhat complicated by the fact that warning lines are only necessarily accurate at the paragraph level, requiring some additional processing. After much delay and burnout, I've finally gotten around to implementing it myself with this PR.

As a bonus, the script now works equally as well locally as remotely; just pass the commits/branches/tags to compare with `--check-and-annotate-changes <base> <branch>`, or use the default (comparing the changes on the current `HEAD` with the merge base in the local `main`). Handy!

* Issue: gh-101100

----
:books: Documentation preview :books:: https://cpython-previews--108065.org.readthedocs.build/

### CODE REVIEW

File: .github/workflows/reusable-docs.yml
Comment (AA-Turner) [inline .github/workflows/reusable-docs.yml]: Will we always need to check out the full history? (Is there some sensible mid-point, or no?)

Note that this takes ~10x longer than the default:
* https://github.com/python/cpython/actions/runs/5884128575/job/15958244549
* https://github.com/python/cpython/actions/runs/5885066149/job/15960917043

File: Doc/library/datetime.rst
Comment (AA-Turner) [inline Doc/library/datetime.rst:1]: Were these changes intentional?

File: Doc/tools/check-warnings.py
Comment (AA-Turner) [inline Doc/tools/check-warnings.py:52]:
```suggestion
```

File: Doc/tools/check-warnings.py
Comment (AA-Turner) [inline Doc/tools/check-warnings.py:56]: For Windows:
```suggestion
        text=True,
        encoding="utf-8",
```

File: Doc/tools/check-warnings.py
Comment (AA-Turner) [inline Doc/tools/check-warnings.py:54]:
```suggestion
        capture_output=True,
```

File: Doc/tools/check-warnings.py
Comment (AA-Turner) [inline Doc/tools/check-warnings.py:61]: Does git return absolute paths?
```suggestion
    return {Path(file.strip()).resolve() for file in added_files if file.strip()}
```

File: Doc/tools/check-warnings.py
Comment (AA-Turner) [inline Doc/tools/check-warnings.py:77]:
```suggestion
        check_output=True,
        check=True,
        text=True,
        encoding="utf-8",
```

File: Doc/tools/check-warnings.py
Comment (AA-Turner) [inline Doc/tools/check-warnings.py]: Perhaps add a comment explaining the diff filter used?
```suggestion
            # https://git-scm.com/docs/git-diff#Documentation/git-diff.txt---diff-filterACDMRTUXB82308203
            "--diff-filter=M",
```

File: Doc/tools/check-warnings.py
Comment (AA-Turner) [inline Doc/tools/check-warnings.py]: I found this a little hard to follow initially -- it could be extracted into a helper function, though may not be worth it.

**Review by AA-Turner — COMMENTED** [review]
Thank you for this work! Some initial comments:

File: .github/workflows/reusable-docs.yml
Comment (CAM-Gerlach) [reply .github/workflows/reusable-docs.yml]: We _should_ only need the history of both branches up to and including the common merge-base. However, unfortunately that's non-trivial to retrieve without hacky, non-trivial and potentially unreliable workarounds. See actions/checkout#552 for more discussion of this. 

While ideally I'd like to avoid this, in the end its an extra 1 minute (30%) to the build time of one of our docs jobs, which brings it to being about as long as the RTD preview build (≈4 minutes each), which is what most people probably care about anyway, so in practice the impact is limited. We could potentially explore workarounds here, but that's probably best left for a followup PR.

File: Doc/library/datetime.rst
Comment (CAM-Gerlach) [reply Doc/library/datetime.rst:1]: Yes; when you reviewed I was still in the middle of actually testing whether the annotations worked on GitHub and debugging an issue with it, which is why the PR was marked as "Draft" rather than "Ready for Review", and why I was still amending my existing commits. I've since completed testing and dropped these changes.

File: Doc/tools/check-warnings.py
Comment (CAM-Gerlach) [reply Doc/tools/check-warnings.py:52]: How come? This avoids any ambiguity with refs vs. files and results in a more specific error message is a ref isn't found, which happened multiple times in testing and was what explicitly motivated me adding this.

File: Doc/tools/check-warnings.py
Comment (CAM-Gerlach) [reply Doc/tools/check-warnings.py:56]: Oops—good catch, thanks. I'm normally very diligent to declare encodings and fixed missing encoding declarations several other places in this file, but I somehow completely missed it here. I included the change in my last batch of fixes when rebasing out the doc changes before marking this ready for review.

File: Doc/tools/check-warnings.py
Comment (CAM-Gerlach) [reply Doc/tools/check-warnings.py:54]: I initially had this, but explicitly chose to adopt the current form instead so that `stderr` still prints and is not suppressed on error, which made debugging multiple issues much more difficult.

File: Doc/tools/check-warnings.py
Comment (CAM-Gerlach) [reply Doc/tools/check-warnings.py:61]: No, but that's intentional, because (like the previous version) we're only using `in` checks on them, and it effectively avoids the need for a couple extra steps parsing and normalizing both paths and possible edge cases with the same, as well as makes the `sphinx-warnings.txt` portable across repos/worktrees and locally/CI, both of which I took advantage of when developing, testing and trying this out.

File: Doc/tools/check-warnings.py
Comment (CAM-Gerlach) [reply Doc/tools/check-warnings.py:77]: * As mentioned above, I added the `encoding` arg in my cleanup prior to my final force-push removing the test changes and marking this ready for review
* There's no `check_output` parameter—I'm guessing you meant `capture_output`—but see above for why I explicitly switched to `stdout` here (to avoid hiding vital error messages)

File: Doc/tools/check-warnings.py
Comment (CAM-Gerlach) [reply Doc/tools/check-warnings.py]: Actually, this option is in fact redundant for this call, as we're explicitly passing in a specific modified file to get the diffs on (it was a holdover from the other function, where I explicitly describe it), so I simply removed it in my last cleanup commit.

File: Doc/tools/check-warnings.py
Comment (CAM-Gerlach) [reply Doc/tools/check-warnings.py]: Yeah, I agree. I ended up factoring it and the next line into a separate helper with a docstring and comments, and splitting this one call into two steps (first filter, then regex match), which was much cleaner. I also added comments a couple other more-dense places.

File: .github/workflows/reusable-docs.yml
Comment (hugovk) [reply .github/workflows/reusable-docs.yml]: And I recall someone who was unhappy about the length of the RTD build ;)

The CPython history is huge: 33 years and 120k commits. Could we limit depth to 100 commits? 500? 1,000?

Are any of the options from https://github.blog/2020-12-21-get-up-to-speed-with-partial-clone-and-shallow-clone/ possible, such as blobless clone?

File: Doc/tools/check-warnings.py
Comment (hugovk) [inline Doc/tools/check-warnings.py]:
```suggestion
    """List the files changed between two Git refs, filtered by change type."""
```

File: Doc/tools/check-warnings.py
Comment (hugovk) [inline Doc/tools/check-warnings.py]:
```suggestion
    """List the lines changed between two Git refs for a specific file."""
```

File: Doc/tools/check-warnings.py
Comment (hugovk) [inline Doc/tools/check-warnings.py]: Perhaps call it `*-diff` so we know to give it two `git diff` type args?

```suggestion
        "--check-and-annotate-diff",
```

File: Doc/tools/check-warnings.py
Comment (hugovk) [inline Doc/tools/check-warnings.py]: The `--help` shows:
```
  --check-and-annotate-changed CHECK_AND_ANNOTATE_CHANGED CHECK_AND_ANNOTATE_CHANGED
                        Annotate lines changed between two refs with warnings on GitHub Actions
```

Shall we add some sort of nicer display name?

```suggestion
        nargs=2,
        metavar="commit",
```

```
  --check-and-annotate-changed commit commit
                        Annotate lines changed between two refs with warnings on GitHub Actions
```

**Review by hugovk — COMMENTED** [review]
Thank you very much for this!

A few initial comments.

File: Doc/tools/check-warnings.py
Comment (hugovk) [inline Doc/tools/check-warnings.py]: > As a bonus, the script now works equally as well locally as remotely; just pass the commits/branches/tags to compare with `--check-and-annotate-changes <base> <branch>`, or use the default (comparing the changes on the current `HEAD` with the merge base in the local `main`). Handy!

Yes, sounds handy!

Although the defaults aren't usable via the CLI yet, or should I call it in another way?

```console
❯ python Doc/tools/check-warnings.py
❯ python Doc/tools/check-warnings.py --check-and-annotate-changed
usage: check-warnings.py [-h] [--check-and-annotate-changed CHECK_AND_ANNOTATE_CHANGED CHECK_AND_ANNOTATE_CHANGED] [--fail-if-regression] [--fail-if-improved]
check-warnings.py: error: argument --check-and-annotate-changed: expected 2 arguments
```

File: .github/workflows/reusable-docs.yml
Comment (CAM-Gerlach) [reply .github/workflows/reusable-docs.yml]: > And I recall someone who was unhappy about the length of the RTD build ;)

Hehe, yeah, though that was because it was a long pole on PR checks and the main thing I wanted to look at, so making the regular build run at the same speed is one "solution" to that :joy: 

Mostly, I was just too tired to think more about this after putting way more time and energy into this than I expected yesterday, including debugging issues related to not fetching the the refs we needed here, because it requires such a long iteration cycle (>5 minutes per run) with minimal feedback and no interactivity.

Unfortunately, given:

* There's around 500 commits per month just on `main`
* People often work on branches that are many months or even years old
* If the fork point was earlier than the shallow depth, the CI job will cryptically fail
* AFAIK, because it doesn't change the original fork point, merging with `main` (AFAIK) won't fix this, only rebasing the branch and rewriting history

We'd have to set a pretty high shallow clone depth, of at least >10 000+ to be reasonably unlikely to run into this, and we might still hit it on a long tail of PRs, plus we might still need other workarounds in the code anyway. So it seems we're probably better off either leaving it, or trying to implement the workarounds at the issue linked I cited above. I can take a crack at the latter, especially if I can repro the shallow clone behavior locally, which would make testing, debugging and iterating _way_ easier.

Yeah, a blobless or treeless clone would be a great help here and a big improvement over a "dumb" shallow clone, but unfortunately despite GitHub itself writing that blog post several years ago, there's inexplicably still no support for anything but a bare-bones shallow clone in the official `checkout` action (which, apparently, is required for reusable workflows, which this in fact is); see, e.g. actions/checkout#1152 and many other issues there, so we'll have to hack something together on our own.

File: Doc/tools/check-warnings.py
Comment (CAM-Gerlach) [reply Doc/tools/check-warnings.py]: Sure; I ended up renaming it to the shorter `annotate-diff` since it doesn't actually perform the checks itself, just control whether annotations are emitted between the specified/default refs.

File: Doc/tools/check-warnings.py
Comment (CAM-Gerlach) [reply Doc/tools/check-warnings.py]: Yeah, though it's actually any Git ref (not necessary, and probably not even usually a commit), the lower-case `commit` is hard to parse as a placeholder for me, and we should be clear on the actual semantics of the two different refs, the first being the base ref being compared against and the second being the head ref to compare (which can be done by passing a tuple of strings to `metavar`).

I did have to tweak this a bit to implement your following suggestion (due to `nargs="*"`), but it ends up looking alright, I think:

```text
  --annotate-diff [BASE_REF [HEAD_REF ...]]
                        Add GitHub Actions annotations on the diff for warnings on lines changed between the given refs (main and HEAD, by default)
```

Comment (CAM-Gerlach) [reply Doc/tools/check-warnings.py]: Oops, I forgot to actually implement that. It ended up requiring a change to use `nargs="*"` and a manual length check, since it isn't possible to natively specify that an argument can take a specific variable number of values with argparse (see, e.g. [this SO answer](https://stackoverflow.com/questions/30944976/argparse-requiring-either-2-values-or-none-for-an-optional-argument)), but now it works, so you can pass any of 0, 1 or 2 arguments (just like `git diff`), but no more.

**Comment by CAM-Gerlach** [comment]
(Just to note, I made another commit to continue to test that the annotations appear (and don't appear) as expected on added files, and on added, fixed and modified warnings in modified files. It deliberately fails the new-warnings check, and I also marked it `DO-NOT-MERGE` just to be safe; once this is approved and otherwise ready to merge, I'll drop that commit and then queue it for automerge.)

File: Doc/library/datetime.rst
Comment (AA-Turner) [reply Doc/library/datetime.rst:1]: No worries!

File: Doc/tools/check-warnings.py
Comment (AA-Turner) [reply Doc/tools/check-warnings.py:52]: Didn't realise git used ``--`` to provide better error messages--dropping this

File: Doc/tools/check-warnings.py
Comment (AA-Turner) [inline Doc/tools/check-warnings.py:85]: Perhaps add a comment here to remind the reader which the optional groups are?
```suggestion
        # removed and added are optional groups; default to 1
        line_match.groupdict(default=1) for line_match in line_matches
```

File: Doc/tools/check-warnings.py
Comment (AA-Turner) [inline Doc/tools/check-warnings.py:38]: It seems we don't use ``linea`` and ``removed`` at all, could we simplify the pattern?

File: Doc/tools/check-warnings.py
Comment (AA-Turner) [inline Doc/tools/check-warnings.py:126]:
```suggestion
    non_null_matches = list(filter(None, warning_matches))
```

File: Doc/tools/check-warnings.py
Comment (AA-Turner) [inline Doc/tools/check-warnings.py:7]: Line 245 uses ``list[str] | None``

```suggestion
from __future__ import annotations

import argparse
```

**Review by AA-Turner — COMMENTED** [review]
Some further comments:

File: Doc/tools/check-warnings.py
Comment (CAM-Gerlach) [reply Doc/tools/check-warnings.py:85]:
```suggestion
        # Removed and added line counts are 1 if not printed
        line_match.groupdict(default=1) for line_match in line_matches
```

Added this, with a bit of rewriting to focus more on the meaningful semantics rather than the literal syntax.

File: Doc/tools/check-warnings.py
Comment (CAM-Gerlach) [reply Doc/tools/check-warnings.py:38]: Yeah they aren't used right now, but we might in the future, and I found they made the regex more literate and self-documenting, making it easier to understand and less cryptic. Therefore, I'd kept them for now.

File: Doc/tools/check-warnings.py
Comment (CAM-Gerlach) [reply Doc/tools/check-warnings.py:126]: Is there a benefit here to using the functional constructs here, aside from being a bit shorter? The rest of this script heavily uses the more Pythonic comprehension syntax, so it seems a bit out of place to me to use the functional syntax in this one spot just to save a few characters...

File: Doc/tools/check-warnings.py
Comment (CAM-Gerlach) [reply Doc/tools/check-warnings.py:7]: Since this was originally only meant to run on the CI and is (currently) only on 3.13+ (with backport only theoretically possible to 3.11 at this point), I hadn't added it, but just in case people are running it locally on security-only versions, doesn't hurt to add, I guess.

File: .github/workflows/reusable-docs.yml
Comment (CAM-Gerlach) [reply .github/workflows/reusable-docs.yml]: To make a very long and painful story short, after much testing and experimentation, I managed to adapt a solution from https://github.com/actions/checkout/issues/520#issuecomment-1167205721 to work here for fetching just enough history to compare the two branches, which adds less than a second to the build process, instead of over a minute. In total, at a rough average of ≈50-ish runs per day of this job (out of 125 workflow runs total), it'll pay back the hours I spent on it, at least in CI time, in under a week.

Also, in testing this, I noticed another pre-existing issue present with the previous version: because the workflow runs on the PR ref _merged_ with the base ref, the lines being reported would be out of sync whenever there were upstream changes, since the GitHub annotations and the Files view are based on the line numbers from the PR head ref, not the merge ref. Naturally, I fixed that as well by making the main docs job run on the PR head ref (which also made the shallow checkout more reliable).

**Comment by CAM-Gerlach** [comment]
Since the final changes were just to a comment and a `__future__` import, I've dropped the various test commits to confirm various common and edge cases, so this should be ready to merge pending any final comments/suggestions (which hopefully aren't substantial or I'll have to temporarily add them back, heh).

**Review by AA-Turner — APPROVED** [review]
Thanks again for all the work on this!

A

**Review by hugovk — APPROVED** [review]
Thanks! Do we need to backport this?

File: Doc/tools/check-warnings.py
Comment (CAM-Gerlach) [reply Doc/tools/check-warnings.py:126]: To (over-)indulge my curiosity, microbenchmarking it, seems like the `filter` approach does have a performance advantage, at least on 3.11 ([PEP 709][(https://peps.python.org/pep-0709/) in Python 3.12 is asserted to speed up comprehensions by 2x or more, particularly in cases like this where the ratio of overhead to actual work is maximized). Tested on the full list of nearly 6000 warnings, I got these results:

```ipython
%timeit non_null_matches = [warning for warning in warning_matches if warning]
576 µs ± 28.3 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)

%timeit non_null_matches = list(filter(None, warning_matches))
163 µs ± 3.64 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
```

That's 3.5x faster. On the other hand, in absolute terms this is a sub-millisecond speedup (on my 10 year old laptop) on a line that is only run twice and represents an upper bound theoretical case as in reality it only actually runs on warnings from added and modified files, which is nearly always a small fraction of the total (on the order of tens to hundreds), and will continue to decrease over time. On a more realistic case adding and modifying a few files with a total of 40 warnings, the numbers look like:

```ipython
%timeit non_null_matches = [warning for warning in warning_matches if warning]
4.39 µs ± 210 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)

%timeit non_null_matches = list(filter(None, warning_matches))
1.65 µs ± 23.6 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
```

The relative speedup is the same, but now we're only saving at most couple microseconds per run, which decreases over time due to Python optimizations, improved CI performance and decreased number of warnings. For comparison, that's around 20 million times less than the speedup with fetch/checkout from my changes above, to the point where at a generous estimate of 100 CI runs per day, every day, it would need to run for 550 years to save the same amount of time as a single run with the optimized checkout. Not to mention run for over 10 000 years in the time I spent indulging my curiosity here...but hope it was at least interesting and informative :joy:

Thank you for attending my TED talk, and I leave you with a profound quote:

"Something something something premature optimization is the root of all evil" — Winston Churchill

**Comment by CAM-Gerlach** [comment]
> Do we need to backport this?

I was actually just about to comment on this now as I remembered that I'd forgotten to do so yesterday. The same problems this fixes (warning spam, as well as line numbers being off, etc) will be present in the Files Changed views on the 3.12 branch, where the script is present (its not on 3.11).

We will need to backport PR #106460 first though as it makes comprehensive changes to these same files; its backport was deferred to after the reusable docs changes were merged, which they now are, so it seems we should be able to just go ahead with that backport first, and then this should backport cleanly.

(In case you didn't see it, I posted [this reply](https://github.com/python/cpython/pull/108065#discussion_r1298669881) after the relevant comment chain had already been resolved, which you might find interesting)

**Comment by CAM-Gerlach** [comment]
I had to manually backport #106460 due to some merge conflicts in the `.nitignore`, but it should be good to merge as soon as Thomas gets the chance to review it, and then we can merge and backport this (hopefully cleanly).

**Comment by CAM-Gerlach** [comment]
Alrighty, looks like that backport got in, so I guess we're gogo on this one at last!

**Pull request merged**
